File: typescript/problem_39.ts
type Cell = `${number},${number}`;

function cellKey(r: number, c: number): Cell { return `${r},${c}`; }
function parseCell(k: Cell): [number, number] {
    const [r, c] = k.split(',').map(Number); return [r, c];
}

export function gameOfLifeStep(alive: Set<Cell>): Set<Cell> {
    const counts = new Map<Cell, number>();
    for (const k of alive) {
        const [r, c] = parseCell(k);
        for (let dr = -1; dr <= 1; dr++)
            for (let dc = -1; dc <= 1; dc++) {
                if (dr === 0 && dc === 0) continue;
                const nb = cellKey(r + dr, c + dc);
                counts.set(nb, (counts.get(nb) ?? 0) + 1);
            }
    }
    const next = new Set<Cell>();
    for (const [p, cnt] of counts) {
        if (alive.has(p) && (cnt === 2 || cnt === 3)) next.add(p);
        if (!alive.has(p) && cnt === 3)               next.add(p);
    }
    return next;
}

export function problem39(): void {
    console.log("Problem 39:");
    const blinker = new Set<Cell>([cellKey(0,0), cellKey(0,1), cellKey(0,2)]);
    const s1 = gameOfLifeStep(blinker);
    if (s1.size !== 3) throw new Error("assertion failed");
    const s2 = gameOfLifeStep(s1);
    if (s2.size !== 3) throw new Error("assertion failed");
    if (!s2.has(cellKey(0,0)) || !s2.has(cellKey(0,1)) || !s2.has(cellKey(0,2)))
        throw new Error("assertion failed");
    console.log("All Tests Passed");
}

File: lua/problem_39.lua
local function stepLife(alive)
    local counts = {}
    for k in pairs(alive) do
        local r, c = k:match("(-?%d+),(-?%d+)")
        r, c = tonumber(r), tonumber(c)
        for dr = -1, 1 do
            for dc = -1, 1 do
                if dr ~= 0 or dc ~= 0 then
                    local nk = (r+dr)..",".. (c+dc)
                    counts[nk] = (counts[nk] or 0) + 1
                end
            end
        end
    end
    local next = {}
    for p, cnt in pairs(counts) do
        if alive[p] and (cnt == 2 or cnt == 3) then next[p] = true
        elseif not alive[p] and cnt == 3        then next[p] = true end
    end
    return next
end

local function setSize(t) local n=0; for _ in pairs(t) do n=n+1 end; return n end

local function problem39()
    print("Problem 39:")
    local blinker = {["0,0"]=true, ["0,1"]=true, ["0,2"]=true}
    local s1 = stepLife(blinker)
    assert(setSize(s1) == 3, "assertion failed")
    local s2 = stepLife(s1)
    assert(setSize(s2) == 3, "assertion failed")
    assert(s2["0,0"] and s2["0,1"] and s2["0,2"], "assertion failed")
    print("All Tests Passed")
end

return { problem39 = problem39 }
